### DISCUSSION TITLE
Using PHP 8 Attribute for event_class_map like EventSerializer & EventVersion

### DISCUSSION THREAD

**Discussion by walirazzaq** [spatie/laravel-event-sourcing#315] [Ideas]
Currently the only place to define `event_class_map` is the config. It fine for countable events but as no. of events grows it becomes huge and hard to maintain.
I was thinking if `AnEventClass` has the mapped name on itself using `PHP Attribute` just we have it for `EventVersion`, `EventSerializer` and cache them like EventHandlers

This might scatter the names all over the place and can cause name conflicts for not being able to see in one place, but we can throw an exception when composing the names for duplication. 
Using named events rather than FQCNs decouples the stored events from code structure which is quite helpful. 
Some helper console commands can be there for developer experience to list mapped names or so
I was thinking to pitch similar idea for `Laravel morphMap` but I thought it's almost same and I should get some expert opinion on that 🤔

```PHP
<?php
namespace App\Events;

use Spatie\EventSourcing\StoredEvents\ShouldBeStored;

#[EventName('user_created')]
class UserCreated extends ShouldBeStored 
{
    //...
}
```

**Comment by brendt** [comment]
Not sure whether we want another way of doing the same. I do like your proposal, but ideally it'd be the only way of managing event names in the future, meaning a new major release.

One potential problem is that now you'd need to scan all event classes to build the event map. That shouldn't be an issue if we properly cache the results, but that should be taken into consideration as well.

**Reply by walirazzaq** [reply]
Yes, for small amount of events current way of mapping name is perfectly fine.

One more issue other than scanning and caching could be name conflicts as names will then be part of the event class file hence not in eye view but that can also be tackled while generating cache or introducing another command to quickly check for conflicts.

As you said it can be re-think in major release so more ideas & improvements can then be better analyzed.

Another small thing about aggregate roots event retrievals I am looking at, I would open new discussion for that once thoroughly grasped that

Thanks for you time ♥️
